missioncontrol/openapi/openapi.yaml

```diff
@@ -1260,6 +1501,56 @@ components:
       additionalProperties:
         type: string
 
+    Files:
+      type: array
+      items:
+        "$ref": "#/components/schemas/File"
+
+    File:
+      required:
+        - cid
+        - what
+        - where
+        - start
+      properties:
+        cid:
+          type: string
+        what:
+          type: string
+        where:
+          type: string
+        created:
+          type: string
+          format: date-time
+          readOnly: true
+        start:
+          type: string
+          format: date-time
+        end:
+          type: string
+          format: date-time
+          nullable: true
+        task_run:
+          type: string
+          format: uuid
+          nullable: true
+        post_url_fields:
+          "$ref": "#/components/schemas/PostUrlFields"
+          readOnly: true
+
+
+    PostUrlFields:
```

Comment:
I'm not exactly sure the goal of this one.
Are you thinking the workflow is:
1) hash the file
2) PUT the file metadata
3) get the post_url_fields returned, and upload the file
?
If so, I think you'd want to upload the file before uploading the metadata to avoid race conditions.
If not, can you say more about how this works?

Reply:
I was envisioning those steps yes.
To do an S3 upload from the client directly without going via the `missioncontrol` host, these post_url_fields are required (S3 requires them) so at a minimum you need

1. Get the signed upload request (ie. post_url_fields) from the `missioncontrol` api.
2. Upload the file to s3 directly.

So, if we want to do metadata after the upload we'd have to do

1. hash the file
2. GET the signed upload request from `missioncontrol` api
3. Upload to s3
4. PUT metadata to `missioncontrol`

Which is one more HTTP request. I would have thought putting the metadata first would actually be less prone to race conditions as you won't start an upload until you've successfully written it's metadata.
Feel free to convince me otherwise, the one extra request is not a big deal except for complicating the logic a bit more.

Reply:
The metadata is what makes the file searchable. If the user creates the metadata, but never uploads the file, how will you know? Having the file exist in the bucket but not exposing it doesn't really present any problems (worst case you have some unknown wasted space files in the bucket).
If the metadata exists first, you can't guarantee to the client that any of the files actually exist.

What's your concern about services having creds to upload content directly? (Especially since many cloud providers will allow you to give write access by host).

Reply:
Fair call on the meta vs. file first.
In terms of uploads directly - a publically writable s3 bucket is asking for trouble. Signed requests are the way to go, but that requires a GET from this repo first to get the signed variables.

Reply:
Split out the file upload field request to a separate HTTP GET request.